### PR TITLE
Convert AutoAWQ checkpoints to compressed-tensors

### LINKED ISSUE #2087: [Feature Request][Help Wanted] Convert AutoAWQ checkpoints to compressed-tensors
**Is your feature request related to a problem? Please describe.**
- Add a conversion tool in compressed-tensors which when given an AutoAWQ checkpoint, can convert the model to the compressed-tensors format using the [pack_quantized](https://github.com/vllm-project/compressed-tensors/blob/main/src/compressed_tensors/compressors/quantized_compressors/pack_quantized.py) compressor
- The checkpoint should also contain an updated `quantization_config` in its config.json with metadata about the format. See example: https://huggingface.co/nm-testing/Qwen3-Coder-30B-A3B-Instruct-W4A16-awq/blob/main/config.json
- This can be done by representing the quantization params for the AutoAWQ model using [QuantizationArgs](https://github.com/vllm-project/compressed-tensors/blob/e88e7d4ad6947628c36ac422a7518f86cc5f1bb8/src/compressed_tensors/quantization/quant_args.py#L142) and then applying the [ModelCompressor](https://github.com/vllm-project/compressed-tensors/blob/e88e7d4ad6947628c36ac422a7518f86cc5f1bb8/src/compressed_tensors/compressors/model_compressors/model_compressor.py#L87) to the model

**Describe the solution you'd like**
- A tool which when given an AutoAWQ checkpoint, produced a compressed-tensors formatted model
- The model should be able to run in vLLM without any drop in accuracy

Working implementation will have to account for AutoAWQ's unique ordering for packing 8 int4 weight into a single int32. More detail here:
https://github.com/vllm-project/llm-compressor/issues/1909#issuecomment-3392856822

### PR DESCRIPTION
**Pull request by mutichung**
## Summary

This PR introduces a new script to convert AutoAWQ checkpoints into `compressed-tensors`-compatible format under `modifiers/awq`. Resolves #2087.

## Usage

- Via CLI:
  
  ```sh
  python -m llmcompressor.modifiers.awq.convert_autoawq \
    --model-name-or-path /path/to/model \
    --output-dir /path/to/compressed/model \
    --quantization-format naive-quantized
  ```

- Via Python:

  ```python
  from llmcompressor.modifiers.awq.convert_autoawq import load_and_convert_from_autoawq
  
  awq_model_path = "/path/to/model"  # can also be model_id on huggingface hub
  model = load_and_convert_from_autoawq(awq_model_path)
  ```

### Asymmetric Support in `llm-compressor` & `compressed-tensors`

- `AutoAWQ` with version `GEMM` only supports asymmetric quantization [^1].
  - `AssertionError` will be raised despite setting `zero_point=False`.
- Support for zero-point decompression in `PackedQuantizationCompressor` is a WIP [^2].
- **2025/12/15 Update**: zero-point decompression has been merged in [^3] but reverted shortly after [^4].

## Test Plan

- [ ] Create tests to compare output logits between AutoAWQ-dequantized floating-point model and `llmcompressor`-compressed model with `CompressedLinear`.
  - The logits do not satisfy `torch.testing.assert_close`, potentially due to GEMM kernel's internal precision?
- [x] Run and compare benchmark results between `AutoAWQForCausalLM` and vLLM.
  - Using `compressed-tensors` based on [^3].
- [x] Created tests to compare benchmark results between AutoAWQ and `llmcompressor` checkpoints.

### ruikangliu/DeepSeek-R1-Distill-Qwen-1.5B-quantized.awq-autoawq-w4g128

| Format           | Inference Backend | ARC-Easy | ARC-Challenge |
| ---------------- | ----------------- | --------:| -------------:|
| AutoAWQ          | hf                |   0.6435 |        0.3584 |
| naive-quantized  | hf                |   0.6431 |        0.3584 |
| packed-quantized | hf                |   0.6431 |        0.3584 |
| packed-quantized | vllm              |   0.6427 |        0.3592 |

### AMead10/Llama-3.2-3B-Instruct-AWQ

| Format           | Inference Backend | ARC-Easy | ARC-Challenge |
| ---------------- | ----------------- | --------:| -------------:|
| AutoAWQ          | hf                |   0.7976 |        0.5017 |
| naive-quantized  | hf                |   0.7971 |        0.5026 |
| packed-quantized | hf                |   0.7971 |        0.5026 |
| packed-quantized | vllm              |   0.7976 |        0.5043 |

### fbaldassarri/mistralai_Mistral-7B-Instruct-v0.3-autoawq-int4-gs128-asym

| Format           | Inference Backend | ARC-Easy | ARC-Challenge |
| ---------------- | ----------------- | --------:| -------------:|
| AutoAWQ          | hf                |   0.8641 |        0.6280 |
| naive-quantized  | hf                |   0.8645 |        0.6280 |
| packed-quantized | hf                |   0.8645 |        0.6280 |
| packed-quantized | vllm              |   0.8649 |        0.6280 |

## Future Work

- Support other AutoAWQ versions, e.g., GEMV.
- Set default quantization format to `packed-quantized` once asymmetric decompression is finalized.
- Replace `AutoModelForCausalLM` with a more generalized autoclass.

[^1]: [awq/modules/linear/gemm.py#L187](https://github.com/casper-hansen/AutoAWQ/blob/88e4c76b20755db275574e6a03c83c84ba3bece5/awq/modules/linear/gemm.py#L187)

[^2]: #1704

[^3]: [compressed-tensors@f9e7426](https://github.com/vllm-project/compressed-tensors/commit/f9e7426e651481f8ce8f5f18105a3ab50e6507e3)

[^4]: [compressed-tensors@cf5980d](https://github.com/vllm-project/compressed-tensors/commit/cf5980da671171c732fc8a4fdf953fa30627b2bb)

### CODE REVIEW

File: src/llmcompressor/modifiers/awq/convert_autoawq.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        iweight, izeros = reverse_awq_order(iweight, izeros, bits)
+
+        # overflow checks
+        iweight = torch.bitwise_and(iweight, (2**bits) - 1)
+        izeros = torch.bitwise_and(izeros, (2**bits) - 1)
+
+        # fp16 weights
+        scales_interleaved = scales.repeat_interleave(group_size, dim=0)
+        izeros_interleaved = izeros.repeat_interleave(group_size, dim=0)
+        fweight = (iweight - izeros_interleaved) * scales_interleaved
```
[the remark this answers is not in the document]
Comment (mutichung) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:122]: Incorrect statement. GEMM shapes are

- `scales`: `(in_features // group_size, out_features)`
- `qzeros`: `(in_features // group_size, out_features // 8)`
- `qweight`: `(in_features, out_features // 8)`

File: src/llmcompressor/modifiers/awq/convert_autoawq.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    # of compressed-tensors and to allow correct quant/dequantization.
+    shifted_zero_point = zero_point - 2 ** (num_bits - 1)
+
+    state_dict.update(
+        {
+            f"{prefix}.weight": weight.T,
+            f"{prefix}.weight_scale": scales.T,
+            f"{prefix}.weight_zero_point": shifted_zero_point.T,
+        }
+    )
```
[the remark this answers is not in the document]
Comment (mutichung) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:142]: Incorrect statement. GEMM shapes are

- `scales`: `(in_features // group_size, out_features)`
- `qzeros`: `(in_features // group_size, out_features // 8)`
- `qweight`: `(in_features, out_features // 8)`

File: tests/llmcompressor/modifiers/awq/test_convert_autoawq.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        model_args=f"pretrained={model_name_or_path},dtype=float16",
+        tasks=["arc_challenge", "arc_easy"],
+        num_fewshot=5,
+        batch_size=16,
+    )
+
+    return results
+
+
+def compare_models(model_name_or_path: str):
```
Comment (brian-dellabetta) [inline tests/llmcompressor/modifiers/awq/test_convert_autoawq.py:21]: running lm_eval can be expensive, when comparing models we just want to ensure the logits are the same for a given set of input_ids. So one way to make this cheaper would be

```python
input_ids = tokenizer("Hello my name is", return_tensors="pt").input_ids
orig_logits = orig_model.forward(input_ids=input_ids).logits
new_logits = new_model.forward(input_ids=input_ids).logits

# possible things to compare
print(f"Norm Diff {(orig_logits-new_logits).norm()}")
print(f"Norm MSE {torch.nn.MSELoss()(orig_logits,new_logits).norm()}")
print(f"Norm {orig_logits.norm()}, {new_logits.norm()}")
```

File: src/llmcompressor/modifiers/awq/convert_autoawq.py
Comment (brian-dellabetta) [inline src/llmcompressor/modifiers/awq/convert_autoawq.py:1]: this may be more suitable for a higher scope, for example placing it into `examples/` or `tools/`, or even directly into compressed-tensors as it does not involve any src code in llmcompressor. 

This is the first time we've added a feature like this, just posting here to see what the rest of the team thinks, and we can decide after that.

Comment (HDCharles) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:1]: agree, this is not a modifier and should probably be in tools or maybe utils?

Comment (mutichung) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:1]: Here's my take from a user experience perspective. None of them are strong opinions btw 😆:

- Users with their AutoAWQ checkpoint would naturally visit the document site of `llm-compressor` and look for things related to AWQ. Putting the conversion script under the AWQ modifier module improves discoverability.
- Putting it into some other places (`examples/`, `tools/`) + mentioning it in the AWQ modifier's description should also work but adds an extra layer of maintenance effort if things are subjected to changes.
- `compressed-tensors` does not have any documentation at the moment. Users would have to be redirected from some warning messages, AWQ modifier documentation, or vLLM documentation to be aware of the existence of this tool.

Also, just curious, here's a question regarding the purpose of this tool:

- vLLM seems to support serving AutoAWQ checkpoints [^1]. Why do we need to convert the format? Is vLLM planning on dropping the support and removing the AutoAWQ kernels?
- If so, then mentioning this conversion tool in the vLLM doc and the warning/error message makes the most sense.
- Otherwise, what are reasons and the potential entry points do you expect users to have the need to convert AutoAWQ checkpoints?

[^1]: [vLLM AutoAWQ page](https://docs.vllm.ai/en/latest/features/quantization/auto_awq)

File: tests/llmcompressor/modifiers/awq/test_convert_autoawq.py (hunk 3 above)
Comment (mutichung) [reply tests/llmcompressor/modifiers/awq/test_convert_autoawq.py:21]: Got it. Will work on this and see what I can do.

File: src/llmcompressor/modifiers/awq/convert_autoawq.py
Comment (HDCharles) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:1]: AutoAWQ is no longer maintained. It's be good to simplify formats and code while being able to support legacy usage through conversion

Comment (HDCharles) [reply src/llmcompressor/modifiers/awq/convert_autoawq.py:1]: The document site for AWQ and the place where the core AWQ components are located aren't necessarily the same. I doubt people are going file by file in core components to find what they want. There's also no readme in there and the general modifiers readme one level up doesn't mention AWQ (bad). Searching AWQ usually leads users to examples/AWQ where there is a dedicated README. it feels like this is the expected landing spot. What path do you expect users would use to find modifiers/awq as their first landing point?

**Comment by mutichung** [comment]
@brian-dellabetta Great, sounds reasonable to me! I'm happy to move this PR to `compressed-tensors`. Thank you all for the review and precious feedback.

**Comment by mutichung** [comment]
Closing in favor of [compressed-tensors#531](https://github.com/vllm-project/compressed-tensors/pull/531).

**Comment by brian-dellabetta** [comment]
Thanks @mutichung , we can follow up on the compressed-tensors PR. Appreciate it!
